File: cielo24_Java/cielo24/src/cielo24/options/BaseOptions.java
package cielo24.options;

import java.lang.reflect.Field;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.Date;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import com.google.common.base.Objects;

import cielo24.Utils;
import cielo24.utils.Dictionary;
import cielo24.utils.KeyValuePair;
import cielo24.utils.QueryName;

/* The base class. All of the other option classes inherit from it. */
public abstract class BaseOptions {

	/* Returns a dictionary that contains key-value pairs of options, where key is the Name property
     * of the QueryName attribute assigned to every option and value is the value of the property.
     * Options with null value are not included in the dictionary. */
    public Dictionary<String, String> GetDictionary() throws IllegalArgumentException, IllegalAccessException {
        Dictionary<String, String> queryDictionary = new Dictionary<String, String>();
        Field[] fields = this.getClass().getDeclaredFields();
        for (Field field : fields) {
        	Object value = field.get(this);
            if (value != null) { // If field is null, don't include the key-value pair in the dictioanary
                QueryName key = field.getDeclaredAnnotation(QueryName.class);
                queryDictionary.add(key.value(), this.getStringValue(value));
            }
        }
        return queryDictionary;
    }

    /* Returns a query String representation of options */
    public String toQuery() throws IllegalArgumentException, IllegalAccessException {
        Dictionary<String, String> queryDictionary = this.GetDictionary();
        return Utils.toQuery(queryDictionary);
    }

    /* Sets the property whose QueryName attribute matches the key */
    public void populateFromKeyValuePair(KeyValuePair<String, String> pair) throws IllegalArgumentException, IllegalAccessException {
    	Field[] fields = this.getClass().getDeclaredFields();
        for (Field field : fields) {
        	QueryName key = field.getDeclaredAnnotation(QueryName.class);
            Type type = field.getType();
            if (key.value().equals(pair.key)) {
            	field.set(this, this.getValueFromString(pair.value, type));
				return;
			}
        }
        throw new IllegalArgumentException("Invalid option: " + pair.key); // Fail if property not found
    }

    // Array of Strings in the key=value form
    public void populateFromArray(String[] array) throws IllegalArgumentException, IllegalAccessException {
        for (String s : Objects.firstNonNull(array, new String[0])) {
        	Matcher regex = Pattern.compile("([^?=&]+)(=([^&]*))?").matcher(s);
        	regex.matches();
        	this.populateFromKeyValuePair(new KeyValuePair<String, String>(regex.group(1), regex.group(3)));
        }
    }

    /* Converts String into an object */
    protected Object getValueFromString(String str, Type type) {
    	// Quotes are necessary in json
    	return Utils.deserialize("\"" + str + "\"", type);
    }

    /* Converts 'value' into String based on its type. Precondition: value != null */
    protected String getStringValue(Object value) {
        if (value instanceof ArrayList<?>) {
        	// ArrayLists can contain strings and enums, type does not matter
            return Utils.joinQuoteList((ArrayList<?>)value, ", ");
        } else if (value instanceof char[]) {   // char[] (returned as (a, b))
            return Utils.joinCharArray((char[])value, ", ");
        } else if (value instanceof Date) {     // DateTime (in ISO 8601 format)
        	return Utils.dateFormat.format((Date)value);
        } else {                                // Takes care of the rest: Integer, Boolean, String, URL
            return value.toString();
        }
    }
}

File: cielo24_.NET/Cielo24/Options/BaseOptions.cs
using Cielo24.JSON;
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Text.RegularExpressions;

namespace Cielo24.Options
{
    /* The base class. All of the other option classes inherit from it. */
    public abstract class BaseOptions
    {
        /* Returns a dictionary that contains key-value pairs of options, where key is the Name property
         * of the QueryName attribute assigned to every option and value is the value of the property.
         * Options with null value are not included in the dictionary. */
        public virtual Dictionary<string, string> GetDictionary()
        {
            Dictionary<string, string> queryDictionary = new Dictionary<string, string>();
            PropertyInfo[] properties = this.GetType().GetProperties();
            foreach (PropertyInfo property in properties)
            {
                object value = property.GetValue(this, null);
                if (value != null) // If property is null, don't include the key-value pair in the dictioanary
                {
                    QueryName key = (QueryName)property.GetCustomAttributes(typeof(QueryName), true).First();
                    queryDictionary.Add(key.Name, this.GetStringValue(value));
                }
            }
            return queryDictionary;
        }

        /* Returns a query string representation of options */
        public virtual string ToQuery()
        {
            Dictionary<string, string> queryDictionary = this.GetDictionary();
            return Utils.ToQuery(queryDictionary);
        }

        /* Sets the property whose QueryName attribute matches the key */
        public virtual void PopulateFromKeyValuePair(KeyValuePair<string, string> pair)
        {
            PropertyInfo[] properties = this.GetType().GetProperties();
            foreach (PropertyInfo property in properties)
            {
                QueryName key = (QueryName)property.GetCustomAttributes(typeof(QueryName), true).First();
                Type type = property.PropertyType;
                if (key.Name.Equals(pair.Key))
                {
                    property.SetValue(this, this.GetValueFromString(pair.Value, type), null);
                    return;
                }
            }
            throw new ArgumentException("Invalid Option: " + pair.Key);
        }

        // Array of strings in the key=value form 
        public void PopulateFromArray(string[] array)
        {
            foreach (string s in array ?? new string[0])
            {
                Dictionary<string, string> dictionary = Regex.Matches(s, "([^?=&]+)(=([^&]*))?").Cast<Match>().ToDictionary(x => x.Groups[1].Value, x => x.Groups[3].Value);
                this.PopulateFromKeyValuePair(dictionary.First());
            }
        }

        /* Converts string into an object */
        protected object GetValueFromString(string str, Type type)
        {
            object result = JsonConvert.DeserializeObject("\"" + str + "\"", type); // Quotes are necessary in json
            return result;
        }

        /* Converts 'value' into string based on its type. Precondition: value != null */
        protected string GetStringValue(object value)
        {
            if (value is List<string>)
            {
                return Utils.JoinQuoteList<string>((List<string>)value, ", ");
            }
            else if (value is List<Tag>)
            {
                return Utils.JoinQuoteList<Tag>((List<Tag>)value, ", ");
            }
            else if (value is List<Fidelity>)
            {
                return Utils.JoinQuoteList<Fidelity>((List<Fidelity>)value, ", ");
            }
            else if (value is char[])       // char[] (returned as (a, b))
            {
                return "(" + String.Join(", ", ((char[])value)) + ")";
            }
            else if (value is DateTime)     // DateTime (in ISO 8601 format)
            {
                return ((DateTime)value).ToString("yyyy-MM-ddTHH:mm:ss.fffffffzzz");
            }
            else                            // Takes care of the rest: int, bool, string, Uri
            {
                return value.ToString();
            }
        }
    }
}
